tests/tests_pytorch/loops/test_training_loop.py

```diff
@@ -206,3 +207,20 @@ def test_should_stop_early_stopping_conditions_met(
 
     assert (message in caplog.text) is raise_debug_msg
     assert trainer.fit_loop._can_stop_early is early_stop
+
+
+def test_should_clear_references_to_dataloader():
+    """Test that `Trainer` clears references to any dataloader passed as arguments after training."""
+    model = BoringModel()
+
+    dataset = RandomDataset(32, 8192)
+
+    trainer = Trainer(fast_dev_run=True)
+    prev_size = len(dumps(trainer))
+
+    trainer.fit(model, train_dataloaders=torch.utils.data.DataLoader(dataset))
+    new_size = len(dumps(trainer))
+
+    size_diff_mb = (new_size - prev_size) // (1024**2)
+
+    assert size_diff_mb == 0
```

Comment:
We don't need to convert this to MB. Besides, the div here is rounding down 😄 
```suggestion
    assert new_size - prev_size == 0
```

Reply:
Actually, the rounding down was intended because the difference isn't exactly 0 (it's about 3KB). I only wanted to make sure that the reference to the data was gone, hence the conversion to MB.

To get this down to 0, we should find all the leaks (and there could be quite a few). I'll have a look and see how many I can find.

Reply:
So I've had another look at this, and I think bringing the difference down to 0 isn't reasonable even if it's possible. To have the trainer's size unchanged after calling `fit`, we'd have to reset all its instance variables to what they were after it was initted. It's possible, but in some cases, doing this isn't justified. For example, `self.callbacks` contains three callbacks after `init` and `fit` may add some more, so we'd need to somehow reset it back to those three callbacks during the teardown.

I think the problem here that the difference between the dumped lengths only tells us how much memory increased but doesn't tell us how much memory was "leaked". Perhaps we should only be dereferencing objects passed from outside (such as data and models), and not minor internal stuff like running states, flag, strings, and so on.

I'm sure that there's a module that allows us to track these leaks (maybe even `gc`) but as I'm not familiar with any at the moment, I need to check.